### PR TITLE
add post detail display

### PR DESCRIPTION
**Pull request by aliceriot**
#### What are the relevant tickets?

this is part of #27 

#### What's this PR do?

this PR basically implements the front-end UI for a 'detail' view of a `Post`. This includes a container and new route for that, a little bit of logic around when to fetch the relevant `Channel` and `Post` objects, and so on. I also updated the `Post` factory to reflect the latest thinking about what the API will look like.

#### How should this be manually tested?

From any thing which shows a list of posts you should be able to click on the `comments` link to get to a detail view for the post. If it is a text post, you should see something like this:

textpost

If it's a URL post you should see basically the same thing, just without the blue text box.

You should be able to click around to the frontpage, to the subreddit, etc, and go to other posts (without a full-page refresh) with no problem.

Visiting a post URL directly (e.g. http://localhost:8063/channel/voluptatum/0.1173598155720883) should work as well.

### CODE REVIEW

**Comment by noisecapella** [comment]
If I go to `/` and click one of the posts, I see a page that says `Loading` but nothing more. If I then refresh the page I can see the detail view however

**Comment by aliceriot** [comment]
wuuuut it definitely wasn't doing that before. looking into it.

File: static/js/components/Loading.js [hunk 1]
```diff
@@ -4,8 +4,8 @@ import R from 'ramda';
 
 import type { RestState } from '../flow/restTypes';
 
-const anyProcessing = R.any(R.propEq('processing', true));
-const allLoaded = R.all(R.propEq('loaded', true));
+export const anyProcessing = R.any(R.propEq('processing', true));
```
Comment (noisecapella) [inline static/js/components/Loading.js]: Should these be moved to a utility module instead?

**Comment by aliceriot** [comment]
Ah, found the problem. When you hit hte frontpage or a subreddit the `frontpage` and `postsForChannel` functions return posts, which we set into the `posts` reducer using the `setPost` custom action. The problem is that this action doesn't set `loaded` to be `true` - I was not seeing that behavior because I was starting my navigation on a post detail view, in which case the `actions.posts.get` function has been called, and `loaded` has already been set to `true` in the posts reducer.

anyway, I'll change the action to set that.

**Comment by aliceriot** [comment]
ok that should be fixed

File: static/js/containers/PostPage.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  getMatchParams = () => {
+    const { match: { params }} = this.props;
+    return [params.postID, params.channelName];
+  }
+
+  updateRequirements = () => {
+    const { dispatch, channels, posts } = this.props;
+    const [postID, channelName] = this.getMatchParams();
+
+    if (R.isNil(posts.data.get(postID))) {
```
Comment (noisecapella) [inline static/js/containers/PostPage.js]: Do we need these checks here? It seems like we only get to this function if the request already completed successfully

File: static/js/containers/PostPage.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    if (!anyProcessing(restStates) || allLoaded(restStates)) {
+      this.updateRequirements();
+    }
+  }
+
+  renderContents(postID, posts, channelName, channels) {
+    const post = posts.get(postID);
+    const channel = channels.get(channelName);
+
+    if (R.isNil(channel) || R.isNil(post)) {
```
Comment (noisecapella) [inline static/js/containers/PostPage.js]: Is this check necessary? It seems like `renderContents` only gets called if all requests have finished

File: static/js/containers/PostPage.js (hunk 2 above)
Comment (aliceriot) [inline static/js/containers/PostPage.js]: We will only get to this function if a request for the `Post` API has already been made. The trick thing is that the request in question could have been for a *different* post than the one we're currently interested in. Consider this case:

- navigate to `/channel/foobar/postid`
- navigate back to `/channel/foobar`
- navigate to `/channel/foobar/anotherpostid`

at this point we have already hit the posts API once (to fetch `postid`). when we load the `PostPage` component, to try to view `anotherpostid`, we will therefore already have `loaded === true`, but we can't quite trust it. It seems, in my testing, that there is typically 1 render call on the changed component before the new request (to get `anotherpostid`) kicks off, when we will have the case that `processing` is false, `loaded` is true, and the data for `anotherpostid` hasn't yet been fetch. So I added this check just to make sure the data is actually there before we try to render anything with it.

Does that make sense? I ran into the same issue with navigating around between subreddits when I was working on the frontpage UI, so I added an equivalent check to the `ChannelPage` container then.

File: static/js/containers/PostPage.js (hunk 3 above)
Comment (aliceriot) [inline static/js/containers/PostPage.js]: See the comment above

File: static/js/components/Loading.js (hunk 1 above)
Comment (aliceriot) [inline static/js/components/Loading.js]: yeah, that makes sense. `lib/rest.js` or something.

**Comment by aliceriot** [comment]
@noisecapella comments addressed

File: static/js/containers/PostPage.js (hunk 2 above)
Comment (noisecapella) [inline static/js/containers/PostPage.js]: Maybe we should clear the data when the post or channel unmounts, but I'm fine with this solution for now

**Pull request merged**